src/Bicep.LangServer/Providers/ServiceClientCredentialsProvider.cs

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+using Bicep.Core.Configuration;
+using Bicep.Core.Registry.Auth;
+using Microsoft.Azure.Management.ResourceGraph;
+
+namespace Bicep.LanguageServer.Providers
+{
+    /// <summary>
+    /// Creates <see cref="ClientCredentials"/> that is used for authenticating <see cref="ResourceGraphClient"/> to Azure.
+    /// <see cref="AccessToken"/> used for credentials creation is cached and reused as long as it's valid.
+    /// New <see cref="AccessToken"/> is fetched on expiration.
+    /// </summary>
+    public class ServiceClientCredentialsProvider : IServiceClientCredentialsProvider
```

Comment:
>ServiceClientCredentialsProvider [](http://example.com/codeflow?start=17&length=32)

The classes that deal with service client credentials are adding a lot of complexity (for example you had to do you own token caching logic). All of that seems to be because the `Microsoft.Azure.Management.ResourceGraph` package doesn't support `TokenCredential`.

However, I found this one: [https://www.nuget.org/packages/Azure.ResourceManager.ResourceGraph#dependencies-body-tab](https://www.nuget.org/packages/Azure.ResourceManager.ResourceGraph#dependencies-body-tab)

It seems to use the `ArmClient` class that we use for template specs and it supports `TokenCredential`.

Reply:
@majastrz, I updated to use Azure.ResourceManager.ResourceGraph here- https://github.com/Azure/bicep/commit/2f2ea825c4d1054b2dac53b737e6cced2ecca1bf. It's not as performant as the client in Microsoft.Azure.Management.ResourceGraph. Here's what it looks like:

![AzureResourceGraphEndpoint](https://user-images.githubusercontent.com/30270536/221894742-fede05c5-3043-4e6d-9cae-e134e817f571.gif)

With Microsoft.Azure.Management.ResourceGraph, the first time completion is invoked, it takes a little longer. Subsequent calls look good:

![MicrosoftResourceGraphClient](https://user-images.githubusercontent.com/30270536/221896894-59c9f416-e087-4b11-a10f-972bc2f7fed2.gif)

I am inclined to keep the Microsoft.Azure.Management.ResourceGraph change. Let me know your thoughts.

Thanks!

Reply:
The challenge with that is the custom token caching logic that we'll have to maintain. The delay is because it initially has to acquire the token. Could you just send a dummy request initially to hide the delay?

Reply:
As discussed offline yesterday, put the changes behind feature flag:
https://github.com/Azure/bicep/commit/383543588f2e4b9a33bd6f74748f1a5e6cf1341f

And logged issue to figure out the latency in new SDK:
https://github.com/Azure/azure-sdk-for-net/issues/34646

Reply:
Sounds good.